lib/bolt/project_migrator/modules.rb

```diff
@@ -0,0 +1,222 @@
+      # new Puppetfile, install the modules, and then move any remaining modules
+      # to the new moduledir.
+      #
+      private def migrate_modules_from_puppetfile(config, puppetfile_path, moduledir, modulepath)
+        require 'bolt/puppetfile'
+        require 'bolt/puppetfile/installer'
+
+        # Migrating modules includes prompting. Do nothing if stdin is not a tty.
+        unless $stdin.tty?
+          @outputter.print_migrate_step(
+            "stdin is not a tty, unable to automatically migrate project modules."\
+            "To migrate project modules manually, see http://pup.pt/bolt-modules"
+          )
+          return false
+        end
+
+        begin
+          @outputter.print_migrate_step("Parsing Puppetfile at #{puppetfile_path}")
+          puppetfile = Bolt::Puppetfile.parse(puppetfile_path, skip_unsupported_modules: true)
+        rescue Bolt::Error => e
+          @outputter.print_migrate_step("#{e.message} Skipping module migration.")
+          return false
+        end
+
+        # Prompt for direct dependencies
+        modules = select_modules(puppetfile.modules)
+
+        # Create new Puppetfile object
+        puppetfile = Bolt::Puppetfile.new(modules)
+
+        # Attempt to resolve dependencies
+        begin
+          @outputter.print_message('')
+          @outputter.print_migrate_step("Resolving module dependencies, this may take a moment")
+          puppetfile.resolve
+        rescue Bolt::Error => e
+          @outputter.print_migrate_step("#{e.message} Skipping module migration.")
+          return false
+        end
+
+        migrate_managed_modules(puppetfile, puppetfile_path, moduledir)
+
+        # Move remaining modules to 'modules'
+        consolidate_modules(modulepath)
+
+        # Delete old modules that are now managed
+        delete_modules(modulepath.first, puppetfile.modules)
+
+        # Add modules to project
+        update_project_config(modules.map(&:to_hash), config)
+      end
+
+      # Migrates the managed modules. If modules were selected to be managed,
+      # the Puppetfile is rewritten and modules are installed. If no modules
+      # were selected, the Puppetfile is deleted.
+      #
+      private def migrate_managed_modules(puppetfile, puppetfile_path, moduledir)
+        if puppetfile.modules.any?
+          # Show the new Puppetfile content
+          message  = "Generated new Puppetfile content:\n\n"
+          message += puppetfile.modules.map(&:to_spec).join("\n").to_s
+          @outputter.print_migrate_step(message)
+
+          # Write Puppetfile
+          @outputter.print_migrate_step("Updating Puppetfile at #{puppetfile_path}")
+          puppetfile.write(puppetfile_path)
+
+          # Install Puppetfile
+          @outputter.print_migrate_step("Syncing modules from #{puppetfile_path} to #{moduledir}")
+          Bolt::Puppetfile::Installer.new({}).install(puppetfile_path, moduledir)
```

Comment:
Should we actually be threading config through here?

Reply:
The puppetfile config? No. `puppetfile-resolver` only handles resolving from the Puppet Forge. We shouldn't install modules from any other Forge.